src/resources/spi_flash_esp32.cc

```diff
@@ -0,0 +1,285 @@
+    // nand flash
+    if (_nand_flash_device) esp_vfs_fat_nand_unmount(_mount_point, _nand_flash_device);
+    if (_nand_flash_device) spi_nand_flash_deinit_device(_nand_flash_device);
+    _nand_flash_device = null;
+    if (_nand_spi_device) spi_bus_remove_device(_nand_spi_device);
+    _nand_spi_device = null;
+#endif
+    free((void *)_mount_point);
+  }
+
+  void close() {
+    tear_down();
+  }
+
+  esp_flash_t* chip() { return _chip; }
+  void set_data_partition(const esp_partition_t* data_partition) { _data_partition = data_partition; }
+  void set_wl_handle(wl_handle_t handle) { _wl_handle = handle; }
+#ifdef CONFIG_SPI_FLASH_NAND_ENABLED
+  void set_nand_flash_device(spi_nand_flash_device_t *nand_flash_device) { _nand_flash_device = nand_flash_device; }
+  void set_nand_spi_device(spi_device_handle_t nand_spi_device) { _nand_spi_device = nand_spi_device; }
+#endif
+  void set_card(sdmmc_card_t *card) { _card = card; }
+  void set_chip(esp_flash_t *chip) { _chip = chip; }
+private:
+  const char *_mount_point;
+  sdmmc_card_t *_card = null;
+  esp_flash_t *_chip = null;
+  const esp_partition_t *_data_partition = null;
+  wl_handle_t _wl_handle = -1;
+#ifdef CONFIG_SPI_FLASH_NAND_ENABLED
+  spi_nand_flash_device_t *_nand_flash_device = null;
+  spi_device_handle_t _nand_spi_device = null;
+#endif
+};
+
+MODULE_IMPLEMENTATION(spi_flash, MODULE_SPI_FLASH);
+
+PRIMITIVE(init_sdcard) {
+  ARGS(cstring, mount_point, SPIResourceGroup, spi_host, int, gpio_cs, int, format_if_mount_failed, int, max_files, int, allocation_unit_size)
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  char* mp = static_cast<char *>(malloc(strlen(mount_point)));
+  if (!mp) MALLOC_FAILED;
+  strcpy(mp, mount_point);
+
+  auto group = _new SPIFlashResourceGroup(process,  mp);
+
+  if (!group) {
+    free(mp);
+    ALLOCATION_FAILED;
+  }
+
+  sdmmc_host_t host = SDSPI_HOST_DEFAULT();
+  host.slot = spi_host->host_device();
+
+  sdspi_device_config_t slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+  slot_config.host_id = static_cast<spi_host_device_t>(host.slot);
+  slot_config.gpio_cs = static_cast<gpio_num_t>(gpio_cs);
+
+  esp_vfs_fat_sdmmc_mount_config_t mount_config = {
+      .format_if_mount_failed = static_cast<bool>(format_if_mount_failed),
+      .max_files = max_files,
+      .allocation_unit_size = static_cast<size_t>(allocation_unit_size)
+  };
+  sdmmc_card_t* card;
+  esp_err_t ret = esp_vfs_fat_sdspi_mount(mount_point, &host, &slot_config, &mount_config, &card);
+  if (ret != ESP_OK) {
+    return Primitive::os_error(ret, process);
+  }
+
+  group->set_card(card);
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(init_nor_flash) {
+  ARGS(cstring, mount_point, SPIResourceGroup, spi_bus, int, gpio_cs,int, frequency, int, format_if_mount_failed, int, max_files, int, allocation_unit_size)
+
+  if (frequency < 0 || frequency > ESP_FLASH_80MHZ) INVALID_ARGUMENT;
```

Comment:
Can you share more of this code? Maybe a static helper to construct the group would be useful?